include/crm/pengine/internal.h

```diff
@@ -529,6 +529,7 @@ void pe_action_set_flag_reason(const char *function, long line, pe_action_t *act
 
 void pe__set_resource_flags_recursive(pe_resource_t *rsc, uint64_t flags);
 void pe__clear_resource_flags_recursive(pe_resource_t *rsc, uint64_t flags);
+void pe__clear_resource_flag_on_all(pe_working_set_t *data_set, uint64_t flag);
```

Comment:
Nitpick, this should be "flags" because it will actually work with more than one flag or'ed together

Reply:
Would this function be more broadly useful if instead of a `pe_working_set_t *`, it took a `GList *` of resources?

Reply:
I don't see any other such uses currently. It would be fine either way.

I noticed some more worms in the can though. Seeing the recursive function above made me wonder why -a/-A doesn't use something like that. The recursive function clears the flag(s) in the resource and all its children, and so on.

The original code does:

1. Clear pe_rsc_allocating on all top-level resources (so we can detect constraint loops, e.g. A colocated with B, B colocated with C, C colocated with A).

2. If the specified resource is marked as pe_rsc_allocating (of course it isn't at the beginning), skip it, otherwise mark it as pe_rsc_allocating. (Previously we would printf "loop" if it was marked -- I don't know when that disappeared.)

3. Loop through all colocation constraints involving the specified resource as the "with"/"right-hand" resource.
3a. If the other resource in the constraint has pe_rsc_allocating, we know we have a loop; skip this constraint.
3b. Otherwise, for -A, repeat from step 2 for the other resource. (***We introduced a bug, we specify rsc instead of peer***)
3c. Output a list item for the constraint.
3d. Print all location constraints for the *other* resource.

4. Print the specified resource and all its location constraints.

5. Clear pe_rsc_allocating on all top-level resources again. (Pre-existing bug: if any of the previously output constraints involved a child resource, it would not have its flag cleared. So I do think we should use pe__clear_resource_flags_recursive() in our new convenience function.)

6. If the specified resource is marked as pe_rsc_allocating (of course it isn't at the beginning), skip it, otherwise mark it as pe_rsc_allocating. (Previously we would printf "loop" if it was marked -- I don't know when that disappeared.)

7. Loop through all colocation constraints involving the specified resource as the "left-hand" resource.
7a. If the other resource in the constraint has pe_rsc_allocating, we know we have a loop; output "loop" and skip this constraint.
7b. Otherwise, output a list item for the constraint.
7c. Print all location constraints for the *other* resource.
7d. For -A, repeat from step 6 for the other resource. (***We introduced a bug, we specify rsc instead of peer***)

Obviously we can fix the bugs, but a bigger question is how the heck this information could be displayed in an intelligible way.

Reply:
For test cases, we can create a single CIB with:

* Nodes N1, N2 (optionally more)
* Resources R1-R10
* Group G with resources GR1-GR3
* Clone C with primitive P
* No constraints for R1
* Colocation chain R2 with R3, R3 with R4, R4 with R5
* Colocation R10 with R4
* Location constraints for R2, R4, and R6 on a node or two each
* Colocation R7 with G
* Colocation R8 with GR2
* Colocation R9 with C

then run crm_resource -a with and without --recursive for each resource R1-R9, G, and C. We could do even more (cloned group, bundle, ...) but I think that would be sufficient.

Reply:
Quickly responding to two easy things in this comment...

> I don't see any other such uses currently. It would be fine either way.

I was just thinking in the future, it could be useful to be able to clear flags on any arbitrary list of resources.  I don't know if anything like that would ever come up, and it's a private function so we can make this change later should we need to, but I could also do it right now.

>     2. If the specified resource is marked as pe_rsc_allocating (of course it isn't at the beginning), skip it, otherwise mark it as pe_rsc_allocating. (Previously we would printf "loop" if it was marked -- I don't know when that disappeared.)

I removed this in a74e1193e005994ec4cfe1991e13bb61fff64de9, which is the commit that did the initial formatted output here.  I could put something back in its place.

Reply:
I imagine something like:

    # crm_resource -a --recursive -r R4
    * Location preferences for R4:
       * Node N1 (score=100, id=I)
       * Node N2 (score=200, id=I)
    * Resources colocated with R4:
       * R3 with R4 (score=N, id=I)
          (here we would output "Location preferences for R3" if there were any)
          * Resources colocated with R3:
             * R2 with R3 (score=N, id=I)
          (here we would somehow output "Resources R3 is colocated with" excluding R4 if there were any) 
       * R10 with R4 (score=N, id=I)
         (here we would output location/colocations for R10 if there were any)
    * Resources R4 is colocated with:
       * R4 with R5 (score=N, id=I)
         (here we would output location/colocations for R5 if there were any)

or maybe more concisely:

    # crm_resource -a --recursive -r R4
    * Location preference for R4: score=100 on N1 (id=I)
    * Location preference for R4: score=200 on N2 (id=I)
    * R3 is colocated with R4 (score=N, id=I)
          * R2 is colocated with R3 (score=N, id=I)
    * R10 is colocated with R4 (score=N, id=I)
    * R4 is colocated with R5 (score=N, id=I)

One challenge I see is that a constraint listing might be a sublist heading or it might just be a list item.

(Also, I missed adding a colocation loop to the test CIB.)

Reply:
Actually I see now that the original code is recursive only in the direction of the colocation, i.e. if R3 is colocated with R4 and we recursively show R4, we show constraints where some resource is colocated with R3, but not constraints where R3 is colocated with some resource other than R4. I suppose we can continue that approach to avoid having to filter the specified resource from the opposite constraints.

Reply:
The idea being we're only showing constraints that involve the specified resource, directly or (if recursive) indirectly. That's why they showed it as a cascade, e.g.

    # crm_resource -a --recursive -r R4
    * Location preferences ...
    * Colocations:
       * R2 with R3
          * R3 with R4
             * R4 with R5
       * R10 with R4

Not sure what is the least confusing approach.

Reply:
Also I'm second-guessing whether we should use --recursive. That currently means "and its children", but we'd be using it as "directly or indirectly". Maybe keep -a/-A and improve the help.

Reply:
Dropped the patch to use --recursive.